examples_tests/0.ImportanceSamplingEnvMaps/fullscreen.frag

```diff
@@ -0,0 +1,669 @@
+    Sphere_Sphere(vec3(2.0,0.0,-1.0),0.5,1u,INVALID_ID_16BIT),
+    Sphere_Sphere(vec3(0.0,0.0,-1.0),0.5,2u,INVALID_ID_16BIT),
+    Sphere_Sphere(vec3(-2.0,0.0,-1.0),0.5,3u,INVALID_ID_16BIT),
+    Sphere_Sphere(vec3(2.0,0.0,1.0),0.5,4u,INVALID_ID_16BIT),
+    Sphere_Sphere(vec3(0.0,0.0,1.0),0.5,4u,INVALID_ID_16BIT),
+    Sphere_Sphere(vec3(-2.0,0.0,1.0),0.5,5u,INVALID_ID_16BIT),
+    Sphere_Sphere( vec3(0.5,1.0,0.5), 0.5, 6u, INVALID_ID_16BIT)
+};
+
+#define LIGHT_COUNT 1
+Light lights[LIGHT_COUNT] = { { vec3(30.0,25.0,15.0), 8u } }; // This doesn't matter if theres only a single env light in the scene
+
+#define BSDF_COUNT 7
+BSDFNode bsdfs[BSDF_COUNT] =
+{
+    {{uvec4(floatBitsToUint(vec3(0.8,0.8,0.8)),DIFFUSE_OP),floatBitsToUint(vec4(0.0,0.0,0.0,0.0))}},
+    {{uvec4(floatBitsToUint(vec3(0.8,0.4,0.4)),DIFFUSE_OP),floatBitsToUint(vec4(0.0,0.0,0.0,0.0))}},
+    {{uvec4(floatBitsToUint(vec3(0.4,0.8,0.4)),DIFFUSE_OP),floatBitsToUint(vec4(0.0,0.0,0.0,0.0))}},
+    {{uvec4(floatBitsToUint(vec3(1.02,1.02,1.3)),CONDUCTOR_OP),floatBitsToUint(vec4(1.0,1.0,2.0,0.0))}},
+    {{uvec4(floatBitsToUint(vec3(1.02,1.3,1.02)),CONDUCTOR_OP),floatBitsToUint(vec4(1.0,2.0,1.0,0.0))}},
+    {{uvec4(floatBitsToUint(vec3(1.02,1.3,1.02)),CONDUCTOR_OP),floatBitsToUint(vec4(1.0,2.0,1.0,0.15))}},
+    {{uvec4(floatBitsToUint(vec3(1.4,1.45,1.5)),DIELECTRIC_OP),floatBitsToUint(vec4(0.0,0.0,0.0,0.0625))}}
+};
+
+float scene_getLightChoicePdf(in Light light)
+{
+    return 1.0/float(LIGHT_COUNT);
+}
+
+uint getBSDFLightIDAndDetermineNormal(out vec3 normal, in uint objectID, in vec3 intersection)
+{
+    Sphere sphere = spheres[objectID];
+    normal = Sphere_getNormal(sphere,intersection);
+    return sphere.bsdfLightIDs;
+}
+
+float nbl_glsl_light_deferred_pdf(in Light light, in Ray_t ray)
+{
+#ifdef IMPORTANCE_SAMPLING
+    const vec3 direction = ray._immutable.direction;
+    const vec2 uv = SampleSphericalMap(direction);
+    return (getLuma(textureLod(envMap, uv, 0.0).rgb))/(2.0*nbl_glsl_PI*nbl_glsl_PI);
+#else
+    return 0.f;
+#endif
+}
+
+vec3 nbl_glsl_light_deferred_eval_and_prob(out float pdf, in Light light, in Ray_t ray)
+{
+    // we don't have to worry about solid angle of the light w.r.t. surface of the light because this function only ever gets called from closestHit routine, so such ray cannot be produced (because lights have no BSDFs here)
+    pdf = scene_getLightChoicePdf(light);
+    pdf *= nbl_glsl_light_deferred_pdf(light, ray);
+    return Light_getRadiance(light, normalize(ray._immutable.direction));
+}
+
+vec3 nbl_glsl_light_generate_and_pdf(out float pdf, out float newRayMaxT, in vec3 origin, in nbl_glsl_AnisotropicViewSurfaceInteraction interaction, in bool isBSDF, in vec3 xi, in uint objectID)
+{   
+#ifdef IMPORTANCE_SAMPLING
+    const vec2 phiPdf = texture(phiPdfLUT, xi.xy).xy;
+    const float theta = texture(thetaLUT, xi.y).x;
+
+    vec3 L = vec3(sin(theta)*cos(phiPdf.x), cos(theta), sin(theta)*sin(phiPdf.x));
```

Comment:
there's a `nbl_glsl_sincos` function of computing sin/cos pairs (you need to make sure the input argument ranges are correct though)

Reply:
done